duffy/legacy/main.py

```diff
@@ -0,0 +1,134 @@
+    contact={"name": "CentOS CI", "email": "ci-sysadmin@centos.org"},
+)
+
+
+@app.post("/Node/get")
+async def get_node(data: NodeGetModel):
+    username, password, dest = obtain_credentials(data.key)
+    response = httpx.post(
+        "%s/api/v1/sessions" % dest,
+        json={
+            "nodes_specs": [
+                {
+                    "type": "physical",
+                    "quantity": data.count,
+                    "distro_type": "centos",
+                    "distro_version": data.ver,
+                },
+            ],
+        },
+        auth=(username, password),
+    )
+    respdict = response.json()
+    if response.status_code == HTTP_201_CREATED:
+        legacy_result = {
+            "ssid": respdict["id"],
+            "hosts": [respdict["nodes"][indx]["hostname"] for indx in respdict["nodes"]],
+        }
+        return legacy_result
+    elif response.status_code == HTTP_422_UNPROCESSABLE_ENTITY:
+        return JSONResponse(status_code=200, content="Failed to allocate nodes")
+    elif (
+        response.status_code == HTTP_401_UNAUTHORIZED or response.status_code == HTTP_403_FORBIDDEN
+    ):
+        return JSONResponse(status_code=403, content={"msg": "Invalid duffy key"})
+    else:
+        return JSONResponse(status_code=200, content="Failed to allocate nodes")
+
+
+@app.post("/Node/done")
+async def return_node_on_completion(data: NodeDoneModel):
+    username, password, dest = obtain_credentials(data.key)
+    response = httpx.put(
+        "%s/api/v1/sessions/%s" % (dest, data.ssid),
+        json={"active": False},
+        auth=(username, password),
+    )
+    if response.status_code == 200:
+        return JSONResponse(status_code=200, content="Done")
+    elif (
+        response.status_code == HTTP_401_UNAUTHORIZED or response.status_code == HTTP_403_FORBIDDEN
+    ):
+        return JSONResponse(status_code=403, content={"msg": "Invalid duffy key"})
+    else:
+        return JSONResponse(status_code=200, content="Failed to return nodes on completion")
+
+
+@app.post("/Node/fail")
+async def return_node_on_failure(data: NodeFailModel):
+    username, password, dest = obtain_credentials(data.key)
+    response = httpx.put(
+        "%s/api/v1/sessions/%s" % (dest, data.ssid),
+        json={"active": False},
+        auth=(username, password),
+    )
+    if response.status_code == 200:
+        return JSONResponse(status_code=200, content="Done")
+    elif (
+        response.status_code == HTTP_401_UNAUTHORIZED or response.status_code == HTTP_403_FORBIDDEN
+    ):
+        return JSONResponse(status_code=403, content={"msg": "Invalid duffy key"})
+    else:
+        return JSONResponse(status_code=200, content="Failed to return nodes on failure")
+
+
+@app.post("/Inventory")
+async def get_all_nodes(data: InventoryRequestModel):
+    username, password, dest = obtain_credentials(data.key)
+    response = httpx.get("%s/api/v1/nodes" % dest, auth=(username, password))
```

Comment:
You saw it coming, didn't you? You knew it from the start that we'd need that so all batteries are included ;P

Reply:
Heh, my crystal ball isn't that good! :wink: I had a hunch that this could be useful, but I thought the odds were even that it wouldn't be used and only served as an example.